File: 0152-maximum-product-subarray/0152-maximum-product-subarray.cpp
class Solution {
public:
    int maxProduct(vector<int>& nums) {
        int maxp = 1;
        int minp = 1;
        int ans = INT_MIN;
        for(int i=0; i<nums.size(); i++) {
            int temp = maxp;
            maxp = max(nums[i], nums[i] >=0 ? maxp * nums[i] : minp * nums[i]);
            minp = min(nums[i], nums[i] >=0 ? minp * nums[i] : temp * nums[i]);
            ans = max(maxp, ans);
        }
        return ans;
    }
};

File: 0152-maximum-product-subarray/0152-maximum-product-subarray.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        maxp = 1
        minp = 1
        ans = -11
        for i in range(len(nums)):
            temp = maxp
            maxp = max(maxp * nums[i], minp * nums[i])
            maxp = max(maxp, nums[i])
            minp = min(temp * nums[i], minp * nums[i])
            minp = min(minp, nums[i])
            ans = max(maxp, ans)
        return ans